stl/inc/shared_mutex

```diff
@@ -227,46 +270,58 @@ public:
 
     shared_lock() noexcept = default;
 
-    _NODISCARD_CTOR_LOCK explicit shared_lock(mutex_type& _Mtx)
+    _NODISCARD_CTOR_LOCK
+    _Acquires_shared_lock_(_Mtx)
+    _Acquires_nonreentrant_lock_(_Mtx)
+    explicit shared_lock(mutex_type& _Mtx)
         : _Pmtx(_STD addressof(_Mtx)), _Owns(true) { // construct with mutex and lock shared
         _Mtx.lock_shared();
     }
 
     shared_lock(mutex_type& _Mtx, defer_lock_t) noexcept
         : _Pmtx(_STD addressof(_Mtx)), _Owns(false) {} // construct with unlocked mutex
 
-    _NODISCARD_CTOR_LOCK shared_lock(mutex_type& _Mtx, try_to_lock_t)
+    _NODISCARD_CTOR_LOCK
+    shared_lock(mutex_type& _Mtx, try_to_lock_t)

```

Comment:
These should probably add a:

```cxx
_When_(this->_Owns, _Acquires_shared_lock_(_Mtx))`
```

Reply:
Added more annotations.
Now I have a false positive one:

```cpp
    std::shared_mutex m;

    m.lock_shared();
    (void) m.try_lock_shared(); // warning C26111: Caller failing to release lock '&m' before calling function 'std::shared_mutex::try_lock_shared'.
```

Do you have an idea how to fix this one ?

Reply:
Hi @cdacamar ! Could you look above?

Reply:
The static analysis warnings are outside my wheelhouse.  @Xazax-hun will know more about this one.

Reply:
Hi!

I am confused. Why is the warning above a false positive? As far as I understand, `std::shared_mutex` is not reentrant and the above error is a true positive.

Reply:
http://eel.is/c++draft/thread.sharedmutex.requirements#general-18

Reply:
Oh.
I see.

`If try_lock is called by a thread that already owns the mutex in any mode (shared or exclusive), the behavior is undefined.`

https://en.cppreference.com/w/cpp/thread/shared_mutex/try_lock

Then the warning is actually a good one!